# Human Gene Expression Matrix Creation 
# Written by Paul Ranum 7/3/2018

#import the gene expression data frame generated by Mandalorion
library(readr)

Gene_Expression_127 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/127_S1_L001-OUTPUT/STEP-ONE/abundance.tsv", 
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_128 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/128_S2_L001-OUTPUT/STEP-ONE/abundance.tsv", 
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_129 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/129_S3_L001-OUTPUT/STEP-ONE/abundance.tsv", 
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_130 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/130_S4_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_131 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/131_S5_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_132 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/132_S6_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_133 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/133_S7_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_134 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/134_S8_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_135 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/135_S9_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_136 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/136_S10_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_137 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/137_S11_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_138 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/138_S12_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_139 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/139_S13_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_140 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/140_S14_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_141 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/141_S15_L001-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_HV1 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_HumanVestibular/Human_Vestibular_1_20170612000_S49_L006-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_HV2 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_HumanVestibular/Human_Vestibular_2_20170612000_S50_L006-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_HV3 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_HumanVestibular/Human_Vestibular_3_20170612000_S51_L006-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)
Gene_Expression_HV4 <- read_delim("/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_HumanVestibular/Human_Vestibular_4_20170612000_S52_L006-OUTPUT/STEP-ONE/abundance.tsv",
                                  "\t", escape_double = FALSE, col_types = cols(eff_length = col_number(), 
                                                                                est_counts = col_number(), length = col_number(), 
                                                                                tpm = col_number()), trim_ws = TRUE)

# Set the columnames
colnames(Gene_Expression_127)[5] <- c("Cell_127")

colnames(Gene_Expression_128)[5] <- c("Cell_128")

colnames(Gene_Expression_129)[5] <- c("Cell_129")

colnames(Gene_Expression_130)[5] <- c("Cell_130")

colnames(Gene_Expression_131)[5] <- c("Cell_131")

colnames(Gene_Expression_132)[5] <- c("Cell_132")

colnames(Gene_Expression_133)[5] <- c("Cell_133")

colnames(Gene_Expression_134)[5] <- c("Cell_134")

colnames(Gene_Expression_135)[5] <- c("Cell_135")

colnames(Gene_Expression_136)[5] <- c("Cell_136")

colnames(Gene_Expression_137)[5] <- c("Cell_137")

colnames(Gene_Expression_138)[5] <- c("Cell_138")

colnames(Gene_Expression_139)[5] <- c("Cell_139")

colnames(Gene_Expression_140)[5] <- c("Cell_140")

colnames(Gene_Expression_141)[5] <- c("Cell_141")

colnames(Gene_Expression_HV1)[5] <- c("Cell_HV1")

colnames(Gene_Expression_HV2)[5] <- c("Cell_HV2")

colnames(Gene_Expression_HV3)[5] <- c("Cell_HV3")

colnames(Gene_Expression_HV4)[5] <- c("Cell_HV4")

# Reformat the Ensembl Gene IDs
ensembl_gene_id <- lapply(Gene_Expression_127, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_127 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_128, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_128 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_129, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_129 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_130, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_130 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_131, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_131 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_132, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_132 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_133, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_133 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_134, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_134 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_135, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_135 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_136, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_136 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_137, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_137 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_137, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_137 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_138, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_138 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_139, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_139 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_140, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_140 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_141, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_141 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_HV1, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_HV1 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_HV2, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_HV2 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_HV3, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_HV3 <- ensembl_gene_id
rm(ensembl_gene_id)

ensembl_gene_id <- lapply(Gene_Expression_HV4, function(x) gsub("\\..*", "", x))
ensembl_gene_id <- data.frame(ensembl_gene_id, stringsAsFactors = FALSE)
Gene_Expression_HV4 <- ensembl_gene_id
rm(ensembl_gene_id)

# Merge the data frames to create the single cell expression matrix
Merged_Gene_Expression <- merge(Gene_Expression_127, Gene_Expression_128, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_129, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_130, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_131, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_132, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_133, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_134, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_135, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_136, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_137, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_138, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_139, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_140, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_141, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_HV1, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_HV2, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_HV3, by = "target_id")
Merged_Gene_Expression <- merge(Merged_Gene_Expression, Gene_Expression_HV4, by = "target_id")

Merged_Gene_Expression <- data.frame(Merged_Gene_Expression$target_id, 
                                     Merged_Gene_Expression$Cell_127,
                                     Merged_Gene_Expression$Cell_128,
                                     Merged_Gene_Expression$Cell_129,
                                     Merged_Gene_Expression$Cell_130,
                                     Merged_Gene_Expression$Cell_131,
                                     Merged_Gene_Expression$Cell_132,
                                     Merged_Gene_Expression$Cell_133,
                                     Merged_Gene_Expression$Cell_134,
                                     Merged_Gene_Expression$Cell_135,
                                     Merged_Gene_Expression$Cell_136,
                                     Merged_Gene_Expression$Cell_137,
                                     Merged_Gene_Expression$Cell_138,
                                     Merged_Gene_Expression$Cell_139,
                                     Merged_Gene_Expression$Cell_140,
                                     Merged_Gene_Expression$Cell_141,
                                     Merged_Gene_Expression$Cell_HV1,
                                     Merged_Gene_Expression$Cell_HV2,
                                     Merged_Gene_Expression$Cell_HV3,
                                     Merged_Gene_Expression$Cell_HV4,
                                     stringsAsFactors = FALSE)

colnames(Merged_Gene_Expression) <- c("ensembl_transcript_id",
                                      "127",
                                      "128",
                                      "129",
                                      "130",
                                      "131",
                                      "132",
                                      "133",
                                      "134",
                                      "135",
                                      "136",
                                      "137",
                                      "138",
                                      "139",
                                      "140",
                                      "141",
                                      "HV1",
                                      "HV2",
                                      "HV3",
                                      "HV4")

#Get Human GRCh38 Transcript IDs, Gene IDs, and Gene Names conversion table 
mart_export <- read_delim("/Volumes/MORL/Analysis/ranum/Gene_IDs/mart_export_Human.txt", 
                          ",", escape_double = TRUE, trim_ws = TRUE)
mart_export <- data.frame(mart_export, stringsAsFactors = FALSE)
head(mart_export)
colnames(mart_export) <- c("ensembl_gene_id", "ensembl_transcript_id", "gene_name")

Merged_GeneEx_GeneID <- merge(Merged_Gene_Expression, mart_export, by = "ensembl_transcript_id")


#Take the sum of rows with matching column values to turn transcript level expression into gene expression.
library(plyr)
Merged_GeneEx_GeneID_In <- data.frame(Merged_GeneEx_GeneID, stringsAsFactors = FALSE)
Merged_GeneEx_GeneID_In["ensembl_gene_id"] <- NULL
Merged_GeneEx_GeneID_In["ensembl_transcript_id"] <- NULL


Merged_GeneEx_GeneID_In2 <- lapply(Merged_GeneEx_GeneID_In[,-20], function(Merged_GeneEx_GeneID_In) as.numeric(as.character(Merged_GeneEx_GeneID_In)))
Merged_GeneEx_GeneID_In2 <- data.frame(Merged_GeneEx_GeneID_In2, stringsAsFactors = FALSE)

Merged_GeneEx_GeneID_In2["gene_name"] <- Merged_GeneEx_GeneID_In$gene_name

#Finally sum the expression values for each gene
Merged_GeneEx_GeneID_Out <- aggregate(. ~ gene_name, data=Merged_GeneEx_GeneID_In2, FUN=sum)


saveRDS(Merged_GeneEx_GeneID_Out, file = "/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/R_GeneExpression_Values.RDS")
write.csv(Merged_GeneEx_GeneID_Out, file = "/Volumes/MORL/Analysis/ranum/scRNA-SeqResults/batch_Human_7_02_2018/R_GeneExpression_Values_V2.csv")





